server/report.go

```diff
@@ -0,0 +1,397 @@
+				b.WriteString("\n\n**Skipped the rest of the services.**")
+				return b.String()
+			}
+
+			if abnormalOnly && s.state == okState {
+				continue
+			}
+
+			if !hostWritten {
+				hostWritten = true
+
+				if linesWritten > 0 {
+					b.WriteString("\n\n")
+				}
+
+				b.WriteString(fmt.Sprintf("`%s`:", h))
+				linesWritten++
+
+				if linesWritten == maximumReportLength {
+					continue
+				}
+
+				// This additional newline produces better Markdown, but we want
+				// to write it only if theEnd hasn't been written.
+				b.WriteRune('\n')
+			}
+
+			b.WriteString(fmt.Sprintf("\n- %s `%s` %s", emoji(s.state), s.name, strings.ToUpper(s.state)))
+			linesWritten++
+		}
+	}
+
+	if linesWritten == 0 {
+		b.WriteString("No services to show.")
+	}
+
+	return b.String()
+}
+
+func (p *Plugin) sendMessages(channelID string, messages ...string) error {
+	for _, m := range messages {
+		post := &model.Post{
+			UserId:    p.botUserID,
+			ChannelId: channelID,
+			Message:   m,
+		}
+		if _, err := p.API.CreatePost(post); err != nil {
+			return fmt.Errorf("p.API.CreatePost: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func (p *Plugin) sendMonitoringReport(channelID string) error {
+	hostCountReq := nagios.HostCountRequest{
+		GeneralHostRequest: nagios.GeneralHostRequest{
+			FormatOptions: nagios.FormatOptions{
+				Enumerate: true,
+			},
+		}}
+
+	var hostCount nagios.HostCount
+
+	if err := p.client.Query(hostCountReq, &hostCount); err != nil {
+		return fmt.Errorf("client.Query: %w", err)
+	}
+
+	hostListReq := nagios.HostListRequest{
+		GeneralHostRequest: nagios.GeneralHostRequest{
+			FormatOptions: nagios.FormatOptions{
+				Enumerate: true,
+			},
+		}}
+
+	var hostList nagios.HostList
+
+	if err := p.client.Query(hostListReq, &hostList); err != nil {
+		return fmt.Errorf("client.Query: %w", err)
+	}
+
+	serviceCountReq := nagios.ServiceCountRequest{
+		GeneralServiceRequest: nagios.GeneralServiceRequest{
+			FormatOptions: nagios.FormatOptions{
+				Enumerate: true,
+			},
+		}}
+
+	var serviceCount nagios.ServiceCount
+
+	if err := p.client.Query(serviceCountReq, &serviceCount); err != nil {
+		return fmt.Errorf("client.Query: %w", err)
+	}
+
+	serviceListReq := nagios.ServiceListRequest{
+		GeneralServiceRequest: nagios.GeneralServiceRequest{
+			FormatOptions: nagios.FormatOptions{
+				Enumerate: true,
+			},
+		}}
+
+	var serviceList nagios.ServiceList
+
+	if err := p.client.Query(serviceListReq, &serviceList); err != nil {
+		return fmt.Errorf("client.Query: %w", err)
+	}
+
+	if err := p.sendMessages(
+		channelID,
+		reportPreamble(time.Now()),
+		formatHostCount(hostCount),
+		formatHostList(hostList),
+		formatServiceCount(serviceCount),
+		formatServiceList(serviceList)); err != nil {
+		return fmt.Errorf("p.sendMessages: %w", err)
+	}
+
+	return nil
+}
+
+const reportLockKey = "report-lock"
+
+var reportLock = []byte{0b10, 0b1, 0b11, 0b111}
+
+func (p *Plugin) monitoringReportLoop() {
```

Comment:
In the HA environment this method will be run on every single node. Do we want that?

Reply:
Yes. I wanted to ensure the system monitoring report's continuous delivery even if one of the plugin instances die (in the HA environment). Since it's not enough to run the reporting loop on one node and running the same code on every node will result in too many reports delivered, I have implemented mutual exclusion (using KV store) of report delivery in `monitoringReportLoop`. All nodes launch this function, but only one node enters the so-called critical section.